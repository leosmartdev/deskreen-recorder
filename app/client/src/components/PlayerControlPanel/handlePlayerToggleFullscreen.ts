/* istanbul ignore file */

// IMPORTANT! leave upper blank line so this file is ignored for coverage!!! More on this issue here
// https://github.com/facebook/create-react-app/issues/6106#issuecomment-550076629
import { REACT_PLAYER_WRAPPER_ID } from "../../constants/appConstants";

export default () => {
  const player = document.querySelector(
    `#${REACT_PLAYER_WRAPPER_ID} > video`
  );
  if (!player) return;
  // @ts-ignore
  if (player.requestFullScreen) {
    // @ts-ignore
    player.requestFullScreen();
    // @ts-ignore
  } else if (player.webkitRequestFullScreen) {
    // @ts-ignore
    player.webkitRequestFullScreen();
    // @ts-ignore
  } else if (player.mozRequestFullScreen) {
    // @ts-ignore
    player.mozRequestFullScreen();
    // @ts-ignore
  } else if (player.msRequestFullscreen) {
    // @ts-ignore
    player.msRequestFullscreen();
    // @ts-ignore
  } else if (player.webkitEnterFullscreen) {
    // @ts-ignore
    player.webkitEnterFullscreen(); //for iphone this code worked
  }
};
