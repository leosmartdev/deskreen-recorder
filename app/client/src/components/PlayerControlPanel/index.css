.play-pause-text {
  width: max-content;
}
