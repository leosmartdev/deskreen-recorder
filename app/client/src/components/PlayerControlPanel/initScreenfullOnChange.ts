import screenfull from "screenfull";

export default (setIsFullScreenOn: (_: boolean) => void) => {
  if (!screenfull.isEnabled) return;
  // @ts-ignore
  screenfull.on('change', () => {
    // @ts-ignore
    setIsFullScreenOn(screenfull.isFullscreen);
  });
};
