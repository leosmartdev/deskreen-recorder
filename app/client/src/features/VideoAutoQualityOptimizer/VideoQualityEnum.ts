export enum VideoQuality {
  Q_AUTO = 'Auto',
  Q_25_PERCENT = '25%',
  Q_40_PERCENT = '40%',
  Q_60_PERCENT = '60%',
  Q_80_PERCENT = '80%',
  Q_100_PERCENT = '100%',
}
