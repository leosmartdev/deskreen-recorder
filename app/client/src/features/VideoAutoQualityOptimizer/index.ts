import { COMPARISON_CANVAS_ID } from './../../constants/appConstants';
import pixelmatch from 'pixelmatch';
import { REACT_PLAYER_WRAPPER_ID } from '../../constants/appConstants';
import CanvasNotDefinedError from './errors/CanvasNotDefinedError';
import VideoDimensionsAreWrongError from './errors/VideoDimensionsAreWrongError';
import VideoNotDefinedError from './errors/VideoNotDefinedError';
import ImageDataIsUndefinedError from './errors/ImageDataIsUndefinedError';

export const CANVAS_SCALE_MULTIPLIER = 0.125; // 1/8 of original canvas size, to speed up calculations
export const MISMATCH_PERCENT_THRESHOLD = 0.1;

export default class VideoAutoQualityOptimizer {
  video: undefined | HTMLVideoElement;

  canvas: undefined | HTMLCanvasElement;

  prevFrame: undefined | ImageData;

  largeMismatchFramesCount = 0;

  isRequestedHalfQuality = false;

  goodQualityCallback = () => {};

  halfQualityCallbak = () => {};

  setGoodQualityCallback(callback: () => void) {
    this.goodQualityCallback = callback;
  }

  setHalfQualityCallbak(callback: () => void) {
    this.halfQualityCallbak = callback;
  }

  startOptimizationLoop() {
    this.prepareCanvasAndVideo();
    setInterval(() => {
      try {
        this.doFrameComparisonAndQualityOptimization();
      } catch (e) {
        // some errors may be thrown here, better ignore them in production
        if (process.env.NODE_ENV === 'development') {
          console.error(e);
        }
      }
    }, 1000);
  }

  doFrameComparisonAndQualityOptimization() {
    this.validateBeforeCalculations();
    this.clearCanvas();
    this.scaleCanvas();
    this.drawVideoFrameToCanvas();
    let imageData = this.getImageDataFromCanvas();

    if (!imageData) {
      throw new ImageDataIsUndefinedError();
    }
    if (!this.prevFrame) {
      this.prevFrame = imageData;
      return;
    }

    try {
      const mismatchInPercent = this.getPreviousAndCurrentFrameMismatchInPercent(imageData);
      this.handleFramesMismatch(mismatchInPercent);
    } catch (e) {
      // usually frames size mismatch thrown here, so can be ignored as it happens
      // often when changing sharing window size
      // so logging this error may be not necessary
    }

    this.prevFrame = imageData;
  }

  findAndSetVideoInternalVariable(document: Document) {
    this.video = document.querySelector(
      `#${REACT_PLAYER_WRAPPER_ID} > video`
    ) as HTMLVideoElement;
  }

  findAndSetCanvasInternalVariable(document: Document) {
    this.canvas = document.querySelector(
      `#${COMPARISON_CANVAS_ID}`
    ) as HTMLCanvasElement;
  }

  prepareCanvasAndVideo() {
    setTimeout(() => {
      this.findAndSetVideoInternalVariable(document);
      this.findAndSetCanvasInternalVariable(document);
    }, 1000);
  }

  clearCanvas() {
    this.canvas
      ?.getContext('2d')
      ?.clearRect(0, 0, this.canvas.width, this.canvas.height);
  }

  validateVideoWidthAndHeight() {
    if (this.video?.videoWidth === 0 || this.video?.videoHeight === 0) {
      throw new VideoDimensionsAreWrongError();
    }
  }

  validateBeforeCalculations() {
    this.validateVideoWidthAndHeight();
    this.validateVideoIsDefined();
    this.validateCanvasIsDefined();
  }

  validateVideoIsDefined() {
    if (!this.video) {
      throw new VideoNotDefinedError();
    }
  }

  validateCanvasIsDefined() {
    if (!this.canvas) {
      throw new CanvasNotDefinedError();
    }
  }

  scaleCanvas() {
    if (
      !this.canvas ||
      !this.video
    )
      return;
    this.canvas.width = this.video.videoWidth * CANVAS_SCALE_MULTIPLIER;
    this.canvas.height = this.video.videoHeight * CANVAS_SCALE_MULTIPLIER;
  }

  drawVideoFrameToCanvas() {
    if (!this.video) return;
    this.canvas
      ?.getContext('2d')
      ?.drawImage(this.video, 0, 0, this.canvas.width, this.canvas.height);
  }

  getImageDataFromCanvas() {
    return this.canvas
      ?.getContext('2d')
      ?.getImageData(0, 0, this.canvas.width, this.canvas.height);
  }

  getNumberOfMismatchedPixels(imageData: ImageData) {
    if (!this.canvas || !this.canvas.width || !this.prevFrame) return 0;
    return pixelmatch(
      this.prevFrame.data,
      imageData.data,
      null,
      this.canvas.width,
      this.canvas.height,
      { threshold: 0.1 }
    );
  }

  getPreviousAndCurrentFrameMismatchInPercent(imageData: ImageData) {
    if (!this.canvas) return 0;
    return this.getNumberOfMismatchedPixels(imageData) / (this.canvas.width * this.canvas.height);
  }

  handleFramesMismatch(mismatchInPercent: number) {
    if (mismatchInPercent < 0.1 && this.largeMismatchFramesCount > 0) {
      this.largeMismatchFramesCount -= 1;
    } else if (mismatchInPercent < 0.1 && this.isRequestedHalfQuality) {
      this.largeMismatchFramesCount = 0;
      this.isRequestedHalfQuality = false;
      this.goodQualityCallback();
    } else if (mismatchInPercent >= 0.1 && !this.isRequestedHalfQuality) {
      if (this.largeMismatchFramesCount < 3) {
        this.largeMismatchFramesCount += 1;
      } else {
        this.halfQualityCallbak();
        this.isRequestedHalfQuality = true;
      }
    }
  }


  isLowMismatchPercent(mismatchInPercent: number) {
    return mismatchInPercent < MISMATCH_PERCENT_THRESHOLD;
  }

  isHighMismatchPercent(mismatchInPercent: number) {
    return mismatchInPercent >= MISMATCH_PERCENT_THRESHOLD;
  }
}
