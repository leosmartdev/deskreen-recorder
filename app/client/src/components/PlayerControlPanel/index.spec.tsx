import React from 'react';
import renderer from 'react-test-renderer';
import PlayerControlPanel from '.';
import ScreenSharingSource from '../../features/PeerConnection/ScreenSharingSourceEnum';
import { VideoQuality } from '../../features/VideoAutoQualityOptimizer/VideoQualityEnum';

jest.useFakeTimers();

it('should match exact snapshot', () => {
  const subject = renderer.create(
    <>
      <PlayerControlPanel
        onSwitchChangedCallback={() => {}}
        isPlaying
        isDefaultPlayerTurnedOn
        handleClickFullscreen={() => {}}
        handleClickPlayPause={() => {}}
        setVideoQuality={() => {}}
        selectedVideoQuality={VideoQuality.Q_100_PERCENT}
        screenSharingSourceType={ScreenSharingSource.SCREEN}
        toaster={undefined}
      />
    </>
  );
  expect(subject).toMatchSnapshot();
});
