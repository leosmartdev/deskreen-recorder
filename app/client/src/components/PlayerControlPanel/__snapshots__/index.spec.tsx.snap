// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match exact snapshot 1`] = `
<div
  className="bp3-card bp3-elevation-4"
>
  <div
    className="row middle-xs between-xs"
  >
    <div
      className="col-xs-12 col-md-3"
    >
      <span
        className="bp3-popover-wrapper"
      >
        <span
          className="bp3-popover-target"
          onBlur={[Function]}
          onFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <button
            className="bp3-button bp3-minimal"
            onBlur={[Function]}
            onClick={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="bp3-button-text"
            >
              <div
                className="row middle-xs"
                style={
                  Object {
                    "opacity": "0.75",
                  }
                }
              >
                <div
                  className="col-xs-4"
                >
                  <img
                    alt="logo"
                    height={42}
                    src="http://localhost/logo192.png"
                    width={42}
                  />
                </div>
                <div
                  className="col-xs-8"
                >
                  <h5
                    className="bp3-heading"
                    style={
                      Object {
                        "marginBottom": "0px",
                      }
                    }
                  >
                    Deskreen
                  </h5>
                </div>
              </div>
            </span>
          </button>
        </span>
      </span>
      <span
        className="bp3-popover-wrapper"
      >
        <span
          className="bp3-popover-target"
          onBlur={[Function]}
          onFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <button
            className="bp3-button"
            onBlur={[Function]}
            onClick={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            style={
              Object {
                "borderRadius": "100px",
              }
            }
            tabIndex={0}
            type="button"
          >
            <span
              className="bp3-button-text"
            >
              <div
                className="row start-xs"
              >
                <div
                  className="col-xs"
                >
                  <img
                    alt="heart"
                    height={16}
                    src="red_heart_2764_twemoji_120x120.png"
                    style={
                      Object {
                        "transform": "translateY(2px)",
                      }
                    }
                    width={16}
                  />
                </div>
                <div
                  className="col-xs"
                >
                  <div
                    style={
                      Object {
                        "transform": "translateY(2px) translateX(-5px)",
                        "width": "max-content",
                      }
                    }
                  >
                    <div
                      className=""
                    >
                      Donate
                    </div>
                  </div>
                </div>
              </div>
            </span>
          </button>
        </span>
      </span>
    </div>
    <div
      className="col-xs-12 col-md-5"
    >
      <div
        className="row center-xs"
        style={
          Object {
            "height": "42px",
          }
        }
      >
        <div
          className="row center-xs middle-xs"
          style={
            Object {
              "backgroundColor": "#137CBD",
              "borderRadius": "20px",
              "height": "100%",
              "width": "190px",
            }
          }
        >
          <div
            className="row center-xs middle-xs"
            style={
              Object {
                "width": "100%",
              }
            }
          >
            <button
              className="bp3-button bp3-minimal"
              onBlur={[Function]}
              onClick={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              style={
                Object {
                  "color": "white",
                  "minWidth": "70px",
                  "width": "85px",
                }
              }
              type="button"
            >
              <span
                className="bp3-button-text"
              >
                <div
                  className="row"
                >
                  <div
                    className="col-xs"
                  >
                    <span
                      className="bp3-icon bp3-icon-pause"
                      icon="pause"
                    >
                      <svg
                        data-icon="pause"
                        fill="white"
                        height={16}
                        viewBox="0 0 16 16"
                        width={16}
                      >
                        <desc>
                          pause
                        </desc>
                        <path
                          d="M6 3H4c-.55 0-1 .45-1 1v8c0 .55.45 1 1 1h2c.55 0 1-.45 1-1V4c0-.55-.45-1-1-1zm6 0h-2c-.55 0-1 .45-1 1v8c0 .55.45 1 1 1h2c.55 0 1-.45 1-1V4c0-.55-.45-1-1-1z"
                          fillRule="evenodd"
                        />
                      </svg>
                    </span>
                  </div>
                  <div
                    className="col-xs"
                  >
                    <div
                      className="bp3-text-large play-pause-text"
                    >
                      Pause
                    </div>
                  </div>
                </div>
              </span>
            </button>
            <div
              className="bp3-divider"
              style={
                Object {
                  "borderBottom": "1px solid #ffffffa8",
                  "borderRight": "1px solid #ffffffa8",
                  "height": "20px",
                }
              }
            />
            <span
              className="bp3-popover-wrapper"
            >
              <span
                className="bp3-popover-target"
                onClick={[Function]}
              >
                <span
                  className="bp3-popover-wrapper"
                >
                  <span
                    className="bp3-popover-target"
                    onBlur={[Function]}
                    onFocus={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                  >
                    <button
                      className="bp3-button bp3-minimal"
                      onBlur={[Function]}
                      onKeyDown={[Function]}
                      onKeyUp={[Function]}
                      tabIndex={0}
                      type="button"
                    >
                      <span
                        className="bp3-button-text"
                      >
                        <span
                          className="bp3-icon bp3-icon-cog"
                          icon="cog"
                        >
                          <svg
                            data-icon="cog"
                            fill="white"
                            height={16}
                            viewBox="0 0 16 16"
                            width={16}
                          >
                            <desc>
                              cog
                            </desc>
                            <path
                              d="M15.19 6.39h-1.85c-.11-.37-.27-.71-.45-1.04l1.36-1.36c.31-.31.31-.82 0-1.13l-1.13-1.13a.803.803 0 00-1.13 0l-1.36 1.36c-.33-.17-.67-.33-1.04-.44V.79c0-.44-.36-.8-.8-.8h-1.6c-.44 0-.8.36-.8.8v1.86c-.39.12-.75.28-1.1.47l-1.3-1.3c-.3-.3-.79-.3-1.09 0L1.82 2.91c-.3.3-.3.79 0 1.09l1.3 1.3c-.2.34-.36.7-.48 1.09H.79c-.44 0-.8.36-.8.8v1.6c0 .44.36.8.8.8h1.85c.11.37.27.71.45 1.04l-1.36 1.36c-.31.31-.31.82 0 1.13l1.13 1.13c.31.31.82.31 1.13 0l1.36-1.36c.33.18.67.33 1.04.44v1.86c0 .44.36.8.8.8h1.6c.44 0 .8-.36.8-.8v-1.86c.39-.12.75-.28 1.1-.47l1.3 1.3c.3.3.79.3 1.09 0l1.09-1.09c.3-.3.3-.79 0-1.09l-1.3-1.3c.19-.35.36-.71.48-1.1h1.85c.44 0 .8-.36.8-.8v-1.6a.816.816 0 00-.81-.79zm-7.2 4.6c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3z"
                              fillRule="evenodd"
                            />
                          </svg>
                        </span>
                      </span>
                    </button>
                  </span>
                </span>
              </span>
            </span>
            <div
              className="bp3-divider"
              style={
                Object {
                  "borderBottom": "1px solid #ffffffa8",
                  "borderRight": "1px solid #ffffffa8",
                  "height": "20px",
                }
              }
            />
            <span
              className="bp3-popover-wrapper"
            >
              <span
                className="bp3-popover-target"
                onBlur={[Function]}
                onFocus={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
              >
                <button
                  className="bp3-button bp3-minimal"
                  onBlur={[Function]}
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  tabIndex={0}
                  type="button"
                >
                  <span
                    className="bp3-button-text"
                  >
                    <img
                      alt="fullscreen-toggle"
                      height={16}
                      src="fullscreen_24px.svg"
                      style={
                        Object {
                          "filter": "invert(100%) sepia(100%) saturate(0%) hue-rotate(127deg) brightness(107%) contrast(102%)",
                          "transform": "scale(1.5) translateY(1px)",
                        }
                      }
                      width={16}
                    />
                  </span>
                </button>
              </span>
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      className="col-xs-12 col-md-3"
    >
      <div
        className="row end-xs"
      >
        <div
          className="col-xs-12"
        >
          <label
            className="bp3-control bp3-switch bp3-disabled bp3-inline bp3-align-right"
            style={
              Object {
                "marginBottom": "0px",
              }
            }
          >
            <input
              checked={true}
              disabled={true}
              onChange={[Function]}
              type="checkbox"
            />
            <span
              className="bp3-control-indicator"
            >
              <div
                className="bp3-control-indicator-child"
              >
                <div
                  className="bp3-switch-inner-text"
                >
                  ON
                </div>
              </div>
              <div
                className="bp3-control-indicator-child"
              >
                <div
                  className="bp3-switch-inner-text"
                >
                  ON
                </div>
              </div>
            </span>
            Default Video Player
          </label>
        </div>
      </div>
    </div>
  </div>
</div>
`;
